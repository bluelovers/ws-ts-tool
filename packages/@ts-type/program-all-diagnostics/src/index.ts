/**
 * Created by user on 2022/6/27.
 */
import { Program, EmitResult, getPreEmitDiagnostics, Diagnostic, flattenDiagnosticMessageText } from 'typescript';

export function getAllDiagnostics(program: Program, emitResult: EmitResult)
{
	return getPreEmitDiagnostics(program)
		.concat(emitResult.diagnostics)
}

export function forEachDiagnostics<T>(allDiagnostics: Diagnostic[], cb: (diagnostic: Diagnostic, message: string) => T)
{
	return allDiagnostics.map(diagnostic =>
	{
		let message = flattenDiagnosticMessageText(diagnostic.messageText, "\n");

		if (diagnostic.file)
		{
			const { line, character } = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start!);

			message = `${diagnostic.file.fileName} (${line + 1},${character + 1}): ${message}`
		}

		return cb(diagnostic, message)
	});
}
