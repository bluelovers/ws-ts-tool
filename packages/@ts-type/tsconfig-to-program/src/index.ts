import { ITsconfig } from '@ts-type/package-dts/tsconfig-json';
import { valueFromRecord } from 'value-from-record';
import {
	JsxEmit,
	ModuleKind,
	ModuleResolutionKind,
	NewLineKind,
	ScriptTarget,
	CompilerOptions,
	ImportsNotUsedAsValues, ModuleDetectionKind,
} from 'typescript';

export function tsconfigToCliArgs(compilerOptions: ITsconfig["compilerOptions"]): string[]
{
	let args = Object.entries(compilerOptions)
		.reduce((a, [key, value]) =>
		{

			if (key === 'locale')
			{
				return a
			}

			if (typeof value === 'boolean')
			{
				if (value === true)
				{
					a.push(`--${key}`);
				}
			}
			else
			{
				a.push(`--${key}`);
				a.push(value);
			}

			return a
		}, [])
	;

	return args;
	//return unparse(compilerOptions)
}

/**
 * for `ts.createProgram`
 */
export function tsconfigToProgram<T extends CompilerOptions>(compilerOptions: ITsconfig["compilerOptions"])
{
	return Object.entries(compilerOptions)
		.reduce((a, [key, value]) =>
		{
			let _skip = false;

			switch (key as keyof ITsconfig["compilerOptions"])
			{
				case 'jsx':
					value = valueFromRecord<string>(value, JsxEmit) ?? value
					break;
				case 'module':
					value = valueFromRecord<ModuleKind>(value, ModuleKind) ?? value
					break;
				case 'moduleResolution':

					value = valueFromRecord<ModuleResolutionKind>(value, ModuleResolutionKind) ?? value

					if (value === 'node')
					{
						value = ModuleResolutionKind.NodeJs
					}

					break;
				case 'moduleDetection':
					value = valueFromRecord<ModuleDetectionKind>(value, ModuleDetectionKind) ?? value
					break;
				case 'newLine':

					switch ((value as string)?.toLowerCase?.())
					{
						case 'lf':
							value = NewLineKind.LineFeed
							break;
						case 'crlf':
							value = NewLineKind.CarriageReturnLineFeed
							break;
						default:
							_skip = true;
					}

					break;
				case 'target':
					value = valueFromRecord<string>(value, ScriptTarget) ?? value;
					break;
				case 'incremental':
					_skip = true;
					break;
				case 'importsNotUsedAsValues':
					value = valueFromRecord<string>(value, ImportsNotUsedAsValues) ?? value;
					break;
			}

			if (!_skip)
			{
				// @ts-ignore
				a[key] = value;
			}

			return a
		}, {} as T)
}
