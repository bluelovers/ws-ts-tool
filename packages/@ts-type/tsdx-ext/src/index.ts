import { ITSTypeAndStringLiteral } from 'ts-type/lib/helper/string';

export const enum EnumTsdxFormat
{
	'cjs' = 'cjs',
	'umd' = 'umd',
	'esm' = 'esm',
	'system' = 'system'
}

export type IModuleFormat = ITSTypeAndStringLiteral<EnumTsdxFormat>;

export function defaultFormatOrder()
{
	return [
		EnumTsdxFormat.esm,
		EnumTsdxFormat.cjs,
		EnumTsdxFormat.umd,
	] as const satisfies readonly IModuleFormat[]
}

export function getExtensionsByFormat(currentFormat: IModuleFormat)
{
	return [
		...(currentFormat === EnumTsdxFormat.cjs ? [
			'.cts' as const,
		] as const : currentFormat === EnumTsdxFormat.esm ? [
			'.mts' as const,
		] as const : currentFormat === EnumTsdxFormat.umd ? [
			'.umd.ts' as const,
		] as const : []),
		'.ts' as const,
		'.tsx' as const,
		...((currentFormat === EnumTsdxFormat.cjs || currentFormat === EnumTsdxFormat.esm) ? [] : [
			'.mts' as const,
			'.cts' as const,
		] as const),
		'.jsx' as const,
		...(currentFormat === EnumTsdxFormat.cjs ? [
			'.cjs' as const,
		] as const : currentFormat === EnumTsdxFormat.esm ? [
			'.mjs' as const,
		] as const : currentFormat === EnumTsdxFormat.umd ? [
			'.umd.js' as const,
		] as const : []),
		'.js' as const,
		...((currentFormat === EnumTsdxFormat.cjs || currentFormat === EnumTsdxFormat.esm) ? [] : [
			'.mjs' as const,
			'.cjs' as const,
		] as const),
		'.js' as const,
	] satisfies `.${string}`[]
}

export default getExtensionsByFormat
