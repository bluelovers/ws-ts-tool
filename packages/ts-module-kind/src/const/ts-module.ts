import type { ModuleKind } from 'typescript';

export enum EnumModuleKind
{
	Node12 = 100,
	None = 0,
	CommonJS = 1,
	AMD = 2,
	UMD = 3,
	System = 4,
	ES2015 = 5,
	ES2020 = 6,
	ES2022 = 7,
	ESNext = 99,
	Node16 = 100,
	NodeNext = 199,
}

export type ITsModuleKindIsCJS = ModuleKind.None | EnumModuleKind.None | ModuleKind.CommonJS | EnumModuleKind.CommonJS | ModuleKind.Node16 | EnumModuleKind.Node16;
export const TS_MODULE_KIND_IS_CJS = [0 as ModuleKind.None | EnumModuleKind.None, 1 as ModuleKind.CommonJS | EnumModuleKind.CommonJS, 100 as ModuleKind.Node16 | EnumModuleKind.Node16] as const;

export type ITsModuleKindIsESM = ModuleKind.ES2015 | EnumModuleKind.ES2015 | ModuleKind.ES2020 | EnumModuleKind.ES2020 | ModuleKind.ES2022 | EnumModuleKind.ES2022 | ModuleKind.ESNext | EnumModuleKind.ESNext | ModuleKind.NodeNext | EnumModuleKind.NodeNext;
export const TS_MODULE_KIND_IS_ESM = [5 as ModuleKind.ES2015 | EnumModuleKind.ES2015, 6 as ModuleKind.ES2020 | EnumModuleKind.ES2020, 7 as ModuleKind.ES2022 | EnumModuleKind.ES2022, 99 as ModuleKind.ESNext | EnumModuleKind.ESNext, 199 as ModuleKind.NodeNext | EnumModuleKind.NodeNext] as const;

