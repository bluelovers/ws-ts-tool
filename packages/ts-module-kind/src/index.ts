import type { CompilerOptions, ModuleKind } from 'typescript';
import { strictEqual } from 'assert';
import {
	EnumModuleKind,
	ITsModuleKindIsCJS,
	ITsModuleKindIsESM,
	TS_MODULE_KIND_IS_CJS,
	TS_MODULE_KIND_IS_ESM,
} from './const/ts-module';
import { ITSAndTypeAndStringLiteral } from 'ts-type/lib/helper/string';
import { valueFromRecord } from 'value-from-record';
import { ITSNumberEnumToNumber } from 'ts-type/lib/helper/record/enum';

export const enum EnumJsKind
{
	cjs = 'cjs',
	esm = 'esm',
}

export const enum EnumJsKindExt
{
	cjs = '.cjs',
	esm = '.mjs',
}

export {
	TS_MODULE_KIND_IS_CJS,
	TS_MODULE_KIND_IS_ESM,
}

export type IEnumModuleKind = typeof EnumModuleKind | typeof ModuleKind;
export type IModuleKind = ModuleKind | EnumModuleKind;
export type IModuleKindKeys = keyof typeof EnumModuleKind | keyof typeof ModuleKind;
export type IModuleKindNumber = ITSNumberEnumToNumber<IModuleKind>;
export type IModuleKindInput = IModuleKind | IModuleKindNumber;
export type IModuleKindInputMixin = ITSAndTypeAndStringLiteral<IModuleKindKeys | IModuleKindInput, string | number>;

function _assertNumber<T extends number>(n: unknown | T): asserts n is T
{
	const type = typeof n;
	strictEqual(type, 'number', new TypeError(`should be a number, but got ${type}`));
}

function _assertString<T extends string>(n: unknown | T): asserts n is T
{
	const type = typeof n;
	strictEqual(type, 'string', new TypeError(`should be a string, but got ${type}`));
}

export function tsModuleKindIsCJS(module: IModuleKindInput | number): module is ITsModuleKindIsCJS
{
	_assertNumber(module);
	return TS_MODULE_KIND_IS_CJS.includes(module)
}

export function tsModuleKindIsESM(module: IModuleKindInput | number): module is ITsModuleKindIsESM
{
	_assertNumber(module);
	return TS_MODULE_KIND_IS_ESM.includes(module)
}

export function isModuleKindName(module: IModuleKindInputMixin): module is IModuleKindKeys
{
	return (typeof module === 'string' && module in EnumModuleKind)
}

export function toModuleKindName(module: IModuleKindInputMixin)
{
	let name: IModuleKindKeys;

	if (isModuleKindName(module))
	{
		name = module
	}
	else
	{
		name = EnumModuleKind[handleModuleKindLazy(module)] as IModuleKindKeys
		_assertString(name);
	}

	return name;
}

export function handleModuleKindLazy(module: IModuleKindInputMixin)
{
	if (typeof module === 'string')
	{
		if (/^\d+$/.test(module))
		{
			module = parseInt(module);
		}
		else
		{
			module = valueFromRecord<ModuleKind>(module, EnumModuleKind)
		}
	}

	return module as ModuleKind
}

export function tsModuleKindIsCJSLazy(module: IModuleKindInputMixin): module is ITsModuleKindIsCJS
{
	return tsModuleKindIsCJS(handleModuleKindLazy(module))
}

export function tsModuleKindIsESMLazy(module: IModuleKindInputMixin): module is ITsModuleKindIsESM
{
	return tsModuleKindIsESM(handleModuleKindLazy(module))
}

export function tsModuleKindType(module: IModuleKindInput)
{
	if (tsModuleKindIsCJS(module))
	{
		return EnumJsKind.cjs as const
	}
	else if (tsModuleKindIsESM(module))
	{
		return EnumJsKind.esm as const
	}
}

export function tsModuleKindExt(module: IModuleKindInput)
{
	if (tsModuleKindIsCJS(module))
	{
		return EnumJsKindExt.cjs as const
	}
	else if (tsModuleKindIsESM(module))
	{
		return EnumJsKindExt.esm as const
	}
}

export function tsModuleKindLazy(module: IModuleKindInputMixin)
{
	return tsModuleKindType(handleModuleKindLazy(module))
}

export function getExtensionsByCompilerOptions(options: CompilerOptions & {
	module?: IModuleKindInputMixin
})
{
	const tsExtensions: string[] = ['.ts'];
	const jsExtensions: string[] = [];

	const module: ModuleKind = handleModuleKindLazy(options.module);

	const useESM = tsModuleKindIsESM(module);
	const useCJS = !useESM && tsModuleKindIsCJS(module);

	if (options.jsx) tsExtensions.push('.tsx');
	if (useESM) tsExtensions.push('.mts');
	else if (useCJS) tsExtensions.push('.cts');

	if (options.allowJs)
	{
		jsExtensions.push('.js');
		if (options.jsx) jsExtensions.push('.jsx');
		if (useESM) tsExtensions.push('.mjs');
		else if (useCJS) tsExtensions.push('.cjs');
	}

	return {
		tsExtensions,
		jsExtensions,
		useESM,
		useCJS,
		module,
	}
}

export default getExtensionsByCompilerOptions
