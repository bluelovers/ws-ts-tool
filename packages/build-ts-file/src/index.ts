/**
 * Created by user on 2020/5/27.
 */

// @ts-ignore
import { JsxEmit, ModuleKind, ModuleResolutionKind, NewLineKind, ScriptTarget, createProgram } from 'typescript';
import { sync as crossSpawn } from 'cross-spawn-extra';
import { ITsconfig } from '@ts-type/package-dts/tsconfig-json';
import { dirname, resolve } from 'path';
// @ts-ignore
import unparse from 'yargs-unparser';
// @ts-ignore
import { getCurrentTsconfig, IOptions as IGetCurrentTsconfigOptions } from 'get-current-tsconfig';
import { consoleLogger } from 'debug-color2/logger';
import { Console2 } from 'debug-color2';
import { tsconfigToCliArgs, tsconfigToProgram } from '@ts-type/tsconfig-to-program';
import { forEachDiagnostics, getAllDiagnostics } from '@ts-type/program-all-diagnostics';

export interface IOptions
{
	bin?: string;
	cwd?: string;
	compilerOptions?: ITsconfig["compilerOptions"];
	getCurrentTsconfigOptions?: IGetCurrentTsconfigOptions;
	verbose?: boolean,
	compilerHost?: import("typescript").CompilerHost | ((programCompilerOptions: import("typescript").CompilerOptions) =>
		import("typescript").CompilerHost),
	overwriteCompilerOptions?: ITsconfig["compilerOptions"];
	logger?: Console2,
}

export function handleOptions(files: string | string[], options?: IOptions)
{
	options ??= {};

	let cwd = options.cwd;

	if (!Array.isArray(files))
	{
		files = [files];
	}

	if (!cwd)
	{
		cwd = dirname(resolve(process.cwd(), files[0]))
	}

	let compilerOptions: ITsconfig["compilerOptions"] = options.compilerOptions ?? getCurrentTsconfig({
		...options.getCurrentTsconfigOptions,
		cwd,
	}).compilerOptions;

	if (options.overwriteCompilerOptions)
	{
		compilerOptions = {
			...compilerOptions,
			...options.overwriteCompilerOptions,
		}
	}

	const bin = options.bin || 'tsc';

	return {
		...options,
		files,
		cwd,
		bin,
		compilerOptions,
	} as IOptions & {
		files: string[];
		cwd: string;
		bin: string;
		compilerOptions: ITsconfig["compilerOptions"];
	}
}

export function spawnEmitTsFiles(inputFiles: string | string[], options?: IOptions)
{

	let { cwd, compilerOptions, files, bin } = handleOptions(inputFiles, options)

	let args = tsconfigToCliArgs(compilerOptions);

	//console.dir(compilerOptions)
	//console.dir(args)
	//console.dir(cwd)
	//console.dir(files)

	let cp = crossSpawn(bin, [
		...args,
		`--tsBuildInfoFile`,
		`.`,
		files[0],
	], {
		cwd,
		stdio: 'inherit',
	});

	return cp
}

/**
 * @see https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API
 */
export function emitTsFiles(inputFiles: string | string[], options?: IOptions)
{
	let {
		cwd,
		compilerOptions,
		files,
		logger,
		verbose,
		compilerHost,
	} = handleOptions(inputFiles, options);

	files = files.map(file => resolve(cwd, file));

	const programCompilerOptions = tsconfigToProgram(compilerOptions);

	if (typeof compilerHost === 'function')
	{
		compilerHost = compilerHost(programCompilerOptions);
	}

	const program = createProgram(files, programCompilerOptions, compilerHost);
	const emitResult = program.emit();

	const exitCode = emitResult.emitSkipped ? 1 : 0;

	let print = logger ?? consoleLogger;

	if (exitCode)
	{
		print = print.red;
	}

	if (verbose)
	{
		const allDiagnostics = getAllDiagnostics(program, emitResult);

		forEachDiagnostics(allDiagnostics, (_diagnostic, message) => {
			print.info(`[Diagnostic]`, message);
		});

		print.debug(`[CWD] ${cwd}`);
	}

	if (exitCode)
	{
		print.error(`[Program] Process exiting with code '${exitCode}'.`);
	}

	//console.log(`Process exiting with code '${exitCode}'.`);

	return {
		cwd,
		files,
		exitCode,
		emitResult,
		compilerOptions,
		programCompilerOptions,
		program,
		compilerHost,
	} as {
		cwd: string;
		files: string[];
		exitCode: number;
		emitResult: import("typescript").EmitResult;
		compilerOptions: ITsconfig["compilerOptions"];
		programCompilerOptions: import("typescript").CompilerOptions;
		program: import("typescript").Program;
		compilerHost: import("typescript").CompilerHost;
	}
}

export default emitTsFiles
