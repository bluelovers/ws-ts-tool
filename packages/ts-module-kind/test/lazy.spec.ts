import { ModuleKind } from 'typescript';
import { handleModuleKindLazy, tsModuleKindLazy } from '../src/index';
import { array_unique } from 'array-hyper-unique';
import { EnumModuleKind } from '../src/const/ts-module';

describe(`ModuleKind`, () =>
{

	array_unique([
		...Object.keys(ModuleKind),
		...Object.keys(EnumModuleKind),
	]).forEach(input =>
	{

		test(`${input}`, () =>
		{

			const module = handleModuleKindLazy(input);

			expect({
				input,
				module,
			}).toStrictEqual({
				input,
				module: expect.any(Number)
			});

			let actual = tsModuleKindLazy(input);
			expect({
				input,
				module,
				moduleName: ModuleKind[module],
				actual,
			}).toMatchSnapshot();

		});

	})

});

test(`Node12`, () =>
{
	expect(EnumModuleKind.Node12).toStrictEqual(ModuleKind.Node16)
	expect(EnumModuleKind[EnumModuleKind.Node12]).toStrictEqual(EnumModuleKind[ModuleKind.Node16])
	expect(EnumModuleKind).toMatchObject(ModuleKind)
})
