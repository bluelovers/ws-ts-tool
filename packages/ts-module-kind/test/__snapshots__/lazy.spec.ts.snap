// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModuleKind 0 1`] = `
{
  "actual": "cjs",
  "input": "0",
  "module": 0,
  "moduleName": "None",
}
`;

exports[`ModuleKind 1 1`] = `
{
  "actual": "cjs",
  "input": "1",
  "module": 1,
  "moduleName": "CommonJS",
}
`;

exports[`ModuleKind 2 1`] = `
{
  "actual": undefined,
  "input": "2",
  "module": 2,
  "moduleName": "AMD",
}
`;

exports[`ModuleKind 3 1`] = `
{
  "actual": undefined,
  "input": "3",
  "module": 3,
  "moduleName": "UMD",
}
`;

exports[`ModuleKind 4 1`] = `
{
  "actual": undefined,
  "input": "4",
  "module": 4,
  "moduleName": "System",
}
`;

exports[`ModuleKind 5 1`] = `
{
  "actual": "esm",
  "input": "5",
  "module": 5,
  "moduleName": "ES2015",
}
`;

exports[`ModuleKind 6 1`] = `
{
  "actual": "esm",
  "input": "6",
  "module": 6,
  "moduleName": "ES2020",
}
`;

exports[`ModuleKind 7 1`] = `
{
  "actual": "esm",
  "input": "7",
  "module": 7,
  "moduleName": "ES2022",
}
`;

exports[`ModuleKind 99 1`] = `
{
  "actual": "esm",
  "input": "99",
  "module": 99,
  "moduleName": "ESNext",
}
`;

exports[`ModuleKind 100 1`] = `
{
  "actual": "cjs",
  "input": "100",
  "module": 100,
  "moduleName": "Node16",
}
`;

exports[`ModuleKind 199 1`] = `
{
  "actual": "esm",
  "input": "199",
  "module": 199,
  "moduleName": "NodeNext",
}
`;

exports[`ModuleKind AMD 1`] = `
{
  "actual": undefined,
  "input": "AMD",
  "module": 2,
  "moduleName": "AMD",
}
`;

exports[`ModuleKind CommonJS 1`] = `
{
  "actual": "cjs",
  "input": "CommonJS",
  "module": 1,
  "moduleName": "CommonJS",
}
`;

exports[`ModuleKind ES2015 1`] = `
{
  "actual": "esm",
  "input": "ES2015",
  "module": 5,
  "moduleName": "ES2015",
}
`;

exports[`ModuleKind ES2020 1`] = `
{
  "actual": "esm",
  "input": "ES2020",
  "module": 6,
  "moduleName": "ES2020",
}
`;

exports[`ModuleKind ES2022 1`] = `
{
  "actual": "esm",
  "input": "ES2022",
  "module": 7,
  "moduleName": "ES2022",
}
`;

exports[`ModuleKind ESNext 1`] = `
{
  "actual": "esm",
  "input": "ESNext",
  "module": 99,
  "moduleName": "ESNext",
}
`;

exports[`ModuleKind Node12 1`] = `
{
  "actual": "cjs",
  "input": "Node12",
  "module": 100,
  "moduleName": "Node16",
}
`;

exports[`ModuleKind Node16 1`] = `
{
  "actual": "cjs",
  "input": "Node16",
  "module": 100,
  "moduleName": "Node16",
}
`;

exports[`ModuleKind NodeNext 1`] = `
{
  "actual": "esm",
  "input": "NodeNext",
  "module": 199,
  "moduleName": "NodeNext",
}
`;

exports[`ModuleKind None 1`] = `
{
  "actual": "cjs",
  "input": "None",
  "module": 0,
  "moduleName": "None",
}
`;

exports[`ModuleKind System 1`] = `
{
  "actual": undefined,
  "input": "System",
  "module": 4,
  "moduleName": "System",
}
`;

exports[`ModuleKind UMD 1`] = `
{
  "actual": undefined,
  "input": "UMD",
  "module": 3,
  "moduleName": "UMD",
}
`;
